### PR TITLE
Add guidance on default params vs function overloading

### PR DESCRIPTION
**Pull request by bondydaa**
Adding section about default params vs function overloading as noted in this PR here: https://github.com/Expensify/Auth/pull/5947/#discussion_r716954684

cc @Expensify/engineering thoughts?

### CODE REVIEW

File: cpp.md [hunk 1]
```diff
@@ -173,3 +173,39 @@ object being passed in.
...
+const string myMethod (const string& myParam, const string& anotherParam)
+{
+    ...do stuff...
+}
+```
+
+Instead in the above case we should prefer a default param like so:
+
+```cpp
+const string myMethod (const string& myParam, const string& anotherParam="")
```
Comment (johnmlee101) [inline cpp.md:197]: Clarification, usually these optional values are declared in the header files, so having `{ do stuff }` is a bit misleading.

Comment (tylerkaraszewski) [reply cpp.md:197]: Yes, I think we should clarify that the defaults actually go in the header files, not the cpp files for the style guide.

**Review by robertjchen — APPROVED** [review]

File: cpp.md (hunk 1 above)
Comment (bondydaa) [reply cpp.md]: updated, you can see it rendered here https://github.com/Expensify/Style-Guides/blob/f1217ffb63a544d87c95eb16f3d7a3c5c406642b/cpp.md

**Review by mountiny — APPROVED** [review]

**Review by stitesExpensify — APPROVED** [review]

File: cpp.md [hunk 2]
```diff
@@ -173,3 +173,30 @@ object being passed in.
...
 merely because you don't *intend* to change something, use it when it is so important not to change it that you want
 the compiler to forbid it.
+
+## Default Params vs Function Overloading
+When deciding whether to give a method a default param or to instead implement function overloading you should use default params when the types for the method will be the same and a default is easily set. If you need to allow or use different types then using function overloading should be preferred.
+
+Ex:
+**Bad** Function Overloading
+```cpp
+const string myMethod (const string& myParam);
```
Comment (cead22) [inline cpp.md:183]: Our style guide should follow our style guide :)
- `const string myMethod(const string& myParam);`
- `const string myMethod(const string& myParam, const string& anotherParam);`
- `const string myMethod(const string& myParam, const string& anotherParam = "");`

**Review by cead22 — CHANGES_REQUESTED** [review]

**Review by Julesssss — COMMENTED** [review]
NAB: The other code examples in this Readme are formatted differently, perhaps they should be aligned?  (good and bad examples in a single code block)

I'd argue we should leave as is though, as your formatting is clearer to me.

File: cpp.md (hunk 2 above)
Comment (bondydaa) [reply cpp.md]: updated

**Review by Julesssss — APPROVED** [review]

**Review by cead22 — APPROVED** [review]

**Review by francoisl — APPROVED** [review]

File: cpp.md [hunk 3]
```diff
@@ -173,3 +173,30 @@ object being passed in.
...
+
+Instead in the above case we should prefer a default param like so:
+
+```cpp
+const string myMethod(const string& myParam, const string& anotherParam = "");
+```
+
+**Good** Function Overloading
+```cpp
+string quotedString(int i) {
```
Comment (Jag96) [inline cpp.md:196]: Shouldn't the `{` here go on the next line?

Comment (bondydaa) [reply cpp.md]: updated

**Review by Jag96 — APPROVED** [review]

**Review by yuwenmemon — APPROVED** [review]

**Review by tylerkaraszewski — APPROVED** [review]

**Review by coleaeason — APPROVED** [review]

**Review by MariaHCD — APPROVED** [review]

**Review by flodnv — APPROVED** [review]

**Pull request merged**